hq/app/db/IamRemediationDb.scala

```diff
@@ -62,27 +80,105 @@ class IamRemediationDb(client: AmazonDynamoDB, tableName: String) {
+    val iamRemediationActivityType = iamRemediationActivity.iamRemediationActivityType
+    val iamProblem = iamRemediationActivity.iamProblem
+    val problemCreationDate = iamRemediationActivity.problemCreationDate
+
+    val iamRemediationActivityTypeString = iamRemediationActivityType match {
+      case Warning => "Warning"
+      case FinalWarning => "FinalWarning"
+      case Remediation => "Remediation"
+    }
+
+    val iamProblemString = iamProblem match {
+      case OutdatedCredential => "OutdatedCredential"
+      case PasswordMissingMFA => "PasswordMissingMFA"
+    }
+
+    val item = Map(
+      "id" -> S(s"${awsAcountId}/${username}"),
+      "awsAccountId" -> S(awsAcountId),
+      "username" -> S(username),
+      "dateNotificationSent" -> N(dateNotificationSent.getMillis),
+      "iamRemediationActivityType" -> S(iamRemediationActivityTypeString),
+      "iamProblem" -> S(iamProblemString),
+      "problemCreationDate" -> N(problemCreationDate.getMillis)
+    )
+
+    new PutItemRequest().withTableName(tableName).withItem(item.asJava)
   }
 
   /**
     * Attempts to deserialise a database query result into our case class.
     */
-  private[db] def deserialiseIamRemediationActivity(dbData: Map[String, AttributeValue]): Attempt[IamRemediationActivity] = {
-    ???
+  private[db] def deserialiseIamRemediationActivity(dbData: Map[String, AttributeValue])(implicit ec: ExecutionContext): Attempt[IamRemediationActivity] = {
+    for {
+      awsAccountId <- valueFromDbData(dbData, "awsAccountId", _.getS)
+      username <- valueFromDbData(dbData, "username", _.getS)
+      dateNotificationSent <- valueFromDbData(dbData, "dateNotificationSent", _.getN.toLong)
+      iamRemediationActivityTypeString <- valueFromDbData(dbData, "iamRemediationActivityType", _.getS)
+      iamRemediationActivity <- iamRemediationActivityFromString(iamRemediationActivityTypeString)
+      iamProblemString <- valueFromDbData(dbData, "iamProblem", _.getS)
+      iamProblem <- iamProblemFromString(iamProblemString)
+      problemCreationDate <- valueFromDbData(dbData, "problemCreationDate", _.getN.toLong)
+    } yield {
+      IamRemediationActivity(awsAccountId,
+        username,
+        new DateTime(dateNotificationSent),
+        iamRemediationActivity,
+        iamProblem,
+        new DateTime(problemCreationDate))
+    }
+  }
+
+  private def valueFromDbData[A](dbData: Map[String, AttributeValue], key: String, f: AttributeValue => A): Attempt[A] =
+    Attempt.fromOption(
+      dbData.get(key).flatMap(data => Option(f(data))),
+      Failure(
+        s"The item retrieved from the database with id ${dbData.get("id")} has an invalid attribute with key $key",
+        s"Failed to deserialise database item into an IamRemediationActivity object",
+        500,
+        None).attempt
+    )
+
+  def iamRemediationActivityFromString(str: String): Attempt[IamRemediationActivityType] = {
+    str match {
+      case "Warning" => Attempt.Right(Warning)
```

Comment:
I don't think we want to change this necessarily but to carry on our conversation from the other day about using a partial function for this... I remembered you can use a Map as a partial function so this could be written as:
```
Attempt.fromOption(
  Map(
    "Warning" -> Warning,
    "FinalWarning" -> FinalWarning,
    "Remediation" -> Remediation
  ).lift(str),
  Failure("Could not turn...", "...", 500).attempt
)
```

I think one nice feature of this is that you could choose to then place all your mappings elsewhere and this code would just sub out the inline `Map` for the `val`.

Reply:
ohh that's a very interesting solution to this problem!